types/datum.go

```diff
@@ -1087,8 +1087,9 @@ func ProduceDecWithSpecifiedTp(dec *MyDecimal, tp *FieldType, sc *stmtctx.Statem
 				return nil, errors.Trace(err)
 			}
 			if !dec.IsZero() && frac > decimal && dec.Compare(&old) != 0 {
-				if sc.InInsertStmt {
+				if sc.InInsertStmt || sc.InUpdateOrDeleteStmt {
```

Comment:
Do we need to add a test for `delete` SQL?

Reply:
Delete behavior is different from MySQL:

```
mysql> select * from decimals;
+-----+
| a   |
+-----+
| 201 |
+-----+
1 row in set (0.00 sec)
```

MySQL:

```
mysql> delete from decimals where a =  cast(201.2 as decimal(20,0));
Query OK, 1 row affected (0.00 sec)
```

TiDB:

```
mysql> delete from decimals where a =  cast(201.2 as decimal(20,0));
Query OK, 1 row affected, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+----------------+
| Level   | Code | Message        |
+---------+------+----------------+
| Warning | 1265 | Data Truncated |
+---------+------+----------------+
1 row in set (0.00 sec)
```

Although it's not the same behavior as MySQL, the result in this PR is more acceptable for delete statement. 
@zimulala